### DISCUSSION TITLE
How to determine the coordinate plane of a PDF file?

### DISCUSSION THREAD

**Discussion by Herrifly** [pymupdf/PyMuPDF#3386] [Looking for help] [answered]
My task is to use page.search_for to find text and highlight it. The problem is that the files may have either different rotation or a different coordinate plane, because I need to mark correctly for any rotation and coordinate plane, then it is enough to understand in what form the plane comes from the PDF file. Tell me how to understand this? Matrix multiplication did not help, I have already tried various ways to determine this. As I understand it, pdf has two types of planes, x horizontal, y vertical and vice versa. I need to be able to identify them, because in this case the drawing coordinates change (since I need to select them horizontally with a stripe). My hypothesis is that these coordinate planes depend on whether the PDF file is in portrait format or landscape. I would be grateful for any parting words!
Maybe there is such a thing?

**Comment by JorjMcKie** [comment]
This is a typical Discussions item, so we converted the original issue.
You can find find out all you need about the coordinate system of a page.
* `page.rotation` contains the normalized page rotation as defined by PDF, one of the values 0, 90, 180, 270.
* `page.rect` the page rectangle from a app programmer view. This means top-left point always has coordinates (0, 0), positive x direction from left to right, positive y direction from top to bottom.
* `page.mediabox`/ `page.cropbox` (and other PDF "boxes"). Mediabox is in PDF's coordinate system - (0, 0) is the bottom-left point. All other boxes have been transformed to MuPDF's systematic: (0, 0) is top-left.
* By default, CropBox is the page rectangle with the following possible deviations:
   - top-left might **_not_** be (0, 0)
   - if page.rotation is 90 or 270, then width and height are reversed in comparison to `page.rect`.
* To ease transformations between PyMuPDF and PDF coordinates, there exist the matrices `page.transformation_matrix` (does PDF -> PyMuPDF), `page.rotation_matrix` (the rotation happening PDF -> PuMuPDF), `page.derotation_matrix` (inverse to the previous). All these matrices are invertible, the inverse being accessible via the `~` operator: `page.rotation_matrix = ~page.derotation_matrix`.

**Reply by Herrifly** [reply]
```
                    rects = page.search_for(phrase)
                    rect = rects[0]
               
                    p1 = rect.tl
                    p2 = rect.br
                    rect_new = fitz.Rect(fitz.Point(x1, p1.y), fitz.Point(x2, p2.y))
                    page.add_rect_annot(rect_new)
```

It's my code, It works correctly on one file with any page rotation, on the other it works correctly when ```
rect_new = fitz.Rect(fitz.Point(p1.x, x1), fitz.Point(p2.x, x2))```
Multiplying matrices does not give the desired result, or I could not multiply them in the right order. Now I can't find a law. Thank's for the answer!

**Comment by Herrifly** [comment]
Can we determine the correct orientation of the file without OCR? 
That is, the text can be rotated and you need to understand how to expand it.

**Reply by JorjMcKie** [reply, marked as answer]
What do you mean?
You have `page.rotation`, you have text writing direction if needed via `page.get_text("dict", ...)`. What else do you need?

**Reply by Herrifly** [reply]
Thank's a lot!
